File: python/multigrid.py
from scipy.interpolate import interp1d
import numpy as np
from smart_slice import SmartSlice, cast

class Grid:
    SHAREDMEM = {}
    DOMAINX = {}

    @classmethod
    def preallocate_memory(cls, n, domain):
        x1, x2 = domain
        for m in range(1, n+1):
            cls.SHAREDMEM[(2**m+1,)] = [np.zeros((2**m+1,)) for i in range(3)]
            cls.DOMAINX[(domain, 2**m+1)] = np.linspace(x1, x2, 2**m+1)

    def __init__(self, domain, n, val=None, func=None, shape=None):
        self.domain = domain
        self.n = n

        N = 2**n
        x1 = domain[0]
        x2 = domain[1]

        self.N = N
        self.x = self.DOMAINX[(domain, N+1)]

        self.h = (x2 - x1)/N
        self._memory_borrowed = False
        
        if val is not None:
            if val.shape[0] != N + 1:
                raise Exception('Shape of val is invalid.')
            self.val = val
        elif func is not None:
            self.val = np.array([func(xi) for xi in self.x])
        elif shape is not None:
            self.val = self.SHAREDMEM[(self.N + 1,) + shape].pop()
            self.val.fill(0)
            self._memory_borrowed = True
        else:
            raise Exception('One of val, func, and shape should be specified.')
            
    def __del__(self):
        if self._memory_borrowed:
            self.SHAREDMEM[self.val.shape].append(self.val)

    def rms(self):
        return np.linalg.norm(self.val, ord=2, axis=0)/np.sqrt(self.val.shape[0])

    def fine(self, kind='linear'):
        finer = Grid(self.domain, self.n + 1, shape=self.val.shape[1:])
        finer.val = interp1d(self.x, self.val, kind=kind, axis=0)(finer.x)
        return finer

    def coarsen(self):
        val = self.val
        coarser = Grid(self.domain, self.n - 1, shape=val.shape[1:])
        # Full Weighting Restriction
        coarser.val[0] = val[0]
        coarser.val[-1] = val[-1]

        # coarser.val[1:coarser.N] = (1/4*val[1:2*coarser.N - 1:2] +
        #                             1/2*val[2:2*coarser.N:2] +
        #                             1/4*val[3:2*coarser.N + 1:2])

        i = SmartSlice(2, 2*coarser.N, 2)
        val = cast(val)
        coarser.val[1:coarser.N] = (1/4*val[i-1] +
                                    1/2*val[i] +
                                    1/4*val[i+1])

        return coarser


class BVP:
    def __init__(self, domain, relax_left_func, relax_middle_func, relax_right_func, res_left_func, res_middle_func, res_right_func, src_func, exact_sol_func=None):
        if not domain[0] < domain[1]:
            raise Exception('Invalid domain.')
        self.domain = domain
        self.relax_left_func = relax_left_func
        self.relax_middle_func = relax_middle_func
        self.relax_right_func = relax_right_func
        self.res_left_func = res_left_func
        self.res_middle_func = res_middle_func
        self.res_right_func = res_right_func
        self.src_func = src_func
        self.exact_sol_func = exact_sol_func

class BVPSolver:
    def __init__(self, bvp, n, src_grid=None, num_iter=(1,1,1)):
        if not isinstance(n, int) and n > 0:
            raise Exception('Invalid n.')

        self.bvp = bvp
        self.n = n

        self.src_grid = Grid(bvp.domain, n, func=bvp.src_func) if src_grid is None else src_grid
        self.sol_grid = Grid(bvp.domain, n, shape=self.src_grid.val.shape[1:])
        
        self.num_iter = num_iter

        # Aliases
        self.h = self.sol_grid.h
        self.x = self.sol_grid.x
        self.N = self.sol_grid.N
        self.sol_val = self.sol_grid.val
        self.src_val = self.src_grid.val

    def relax(self):
        # Aliases
        h = self.h
        x = self.x
        N = self.N
        sol_val = self.sol_val
        src_val = self.src_val

        # Red Sweep
        sol_val[0] = self.bvp.relax_left_func(sol_val, src_val, x[0], h)
        for i in range(2, N, 2):
            sol_val[i] = self.bvp.relax_middle_func(sol_val, src_val, x[i], h, i)
        sol_val[N] = self.bvp.relax_right_func(sol_val, src_val, x[N], h)

        # Black Sweep
        for i in range(1, N, 2):
            sol_val[i] = self.bvp.relax_middle_func(sol_val, src_val, x[i], h, i)

    def residual(self):
        # Aliases
        h = self.h
        x = self.x
        N = self.N
        sol_val = self.sol_val
        src_val = self.src_val

        res = np.zeros(sol_val.shape)
        # Residual at the leftmost point
        res[0] = self.bvp.res_left_func(sol_val, src_val, x[0], h)
        # Residual at middle points
        for i in range(1, N):
            res[i] = self.bvp.res_middle_func(sol_val, src_val, x[i], h, i)
        # Residual at the rightmost point
        res[N] = self.bvp.res_left_func(sol_val, src_val, x[N], h)

        return Grid(self.bvp.domain, self.n, val=res)

    def multigrid(self):
        # Pre-Smoothing
        for i in range(self.num_iter[0]):
            self.relax()
        if self.n != 1:
            # Error Correction using Coarse Grid
            for i in range(self.num_iter[1]):
                coarse = type(self)(self.bvp, self.n - 1, src_grid=self.residual().coarsen(), num_iter=self.num_iter)
                coarse.solve()
                self.sol_val += coarse.sol_grid.fine().val

            # Post-Smoothing
            for i in range(self.num_iter[2]):
                self.relax()

    def solve(self):
        self.multigrid()

    def exact_error(self):
        if self.bvp.exact_sol_func is None:
            raise Exception('exact_sol_func is not specified.')
        
        exact_error = np.array([self.bvp.exact_sol_func(xi) for xi in self.sol_grid.x]) - self.sol_val
        return Grid(self.bvp.domain, self.n, val=exact_error)

File: c/multigrid.c
#include "multigrid.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdio.h>

Grid *grid_new(double x1, double x2, int n)
{
    Grid *grid = (Grid*)malloc(sizeof(Grid));
    grid->x1 = x1;
    grid->x2 = x2;
    grid->n = n;
    int N = 1<<n;
    grid->N = N;
    double h = (x2 - x1)/(double)N;
    grid->h = h;
    grid->x = (double*)malloc(sizeof(double)*(N + 1));
    for (int i = 0; i <= N; i++) {
        grid->x[i] = x1 + i*h;
    }
    grid->val = malloc(sizeof(double)*(N + 1));
    return grid;
}

Grid *grid_new_zeros(double x1, double x2, int n)
{
    Grid *grid = grid_new(x1, x2, n);
    grid_set_zero(grid);
    return grid;
}

Grid *grid_new_func(double x1, double x2, int n, double (*func)(double))
{
    Grid *grid = grid_new(x1, x2, n);
    for (int i = 0; i <= grid->N; i++) {
        grid->val[i] = func(grid->x[i]);
    }
    return grid;
}

void grid_delete(Grid *grid)
{
    free(grid->x);
    free(grid->val);
    free(grid);
}

void grid_set_zero(Grid *grid)
{
    memset(grid->val, 0, sizeof(double)*(grid->N + 1));
}

double grid_rms(Grid *grid)
{
    double sum = 0.;
    for (int i = 0; i <= grid->N; i++) {
        sum += grid->val[i]*grid->val[i];
    }
    return sqrt(sum/(double)(1 + grid->N));
}

void grid_fine(Grid *grid, Grid *target)
{
    // Linear Interpolation
    for (int i = 0; i <= target->N; i++) {
        if (i % 2 == 0) {
            target->val[i] = grid->val[i/2];
        } else {
            target->val[i] = grid->val[i/2]/2. + grid->val[i/2 + 1]/2.;
        }
    }
}

void grid_coarsen(Grid *grid, Grid *target)
{
    // Full Weighting Restriction
    target->val[0] = grid->val[0];
    for (int i = 1; i < target->N; i++) {
        target->val[i] = grid->val[2*i - 1]/4. + grid->val[2*i]/2. + grid->val[2*i + 1]/4.;
    
    }
    target->val[target->N] = grid->val[grid->N];
}

void grid_subtract(Grid *a, Grid *b, Grid *target)
{
    for (int i = 0; i <= target->N; i++) {
        target->val[i] = a->val[i] - b->val[i];
    }
}

BVP *bvp_new(
    double x1, double x2,
    double (*relax_left_func)(double*, double*, double, double),
    double (*relax_middle_func)(double*, double*, double, double, int),
    double (*relax_right_func)(double*, double*, double, double),
    double (*res_left_func)(double*, double*, double, double),
    double (*res_middle_func)(double*, double*, double, double, int),
    double (*res_right_func)(double*, double*, double, double),
    double (*src_func)(double),
    double (*exact_sol_func)(double)
)
{
    BVP *bvp = (BVP*)malloc(sizeof(BVP));
    bvp->x1 = x1;
    bvp->x2 = x2;
    bvp->relax_left_func = relax_left_func;
    bvp->relax_middle_func = relax_middle_func;
    bvp->relax_right_func = relax_right_func;
    bvp->res_left_func = res_left_func;
    bvp->res_middle_func = res_middle_func;
    bvp->res_right_func = res_right_func;
    bvp->src_func = src_func;
    bvp->exact_sol_func = exact_sol_func;
    return bvp;
}

void bvp_delete(BVP *bvp)
{
    free(bvp);
}

BVPSolver *bvpsolver_new(BVP *bvp, int num_level, int num_iter1, int num_iter2, int num_iter3)
{
    BVPSolver *solver = (BVPSolver*)malloc(sizeof(BVPSolver));
    solver->bvp = bvp;
    solver->n = num_level;

    for (GridKind i = 0; i < NUM_GK; i++) {
        solver->grid[i] = (Grid**)malloc(sizeof(Grid*)*num_level);
        for (int j = 0; j < num_level - 1; j++) {
            solver->grid[i][j] = grid_new(bvp->x1, bvp->x2, j + 1);
        }
        switch (i) {
            case SRC:
                solver->grid[i][num_level - 1] = grid_new_func(bvp->x1, bvp->x2, num_level, bvp->src_func);
                break;
            case SOL:
                solver->grid[i][num_level - 1] = grid_new_zeros(bvp->x1, bvp->x2, num_level);
                break;
            case RES:
            case ERR:
                solver->grid[i][num_level - 1] = grid_new(bvp->x1, bvp->x2, num_level);
                break;
        }
    }

    solver->num_iter1 = num_iter1;
    solver->num_iter2 = num_iter2;
    solver->num_iter3 = num_iter3;

    return solver;
}

void bvpsolver_delete(BVPSolver *solver)
{
    for (GridKind i = 0; i < NUM_GK; i++) {
        for (int j = 0; j < solver->n; j++) {
            grid_delete(solver->grid[i][j]);
        }
        free(solver->grid[i]);
    }
    free(solver);
}

void bvpsolver_relax(BVPSolver *solver, int level)
{
    // Aliases
    double h = solver->grid[0][level]->h;
    double *x = solver->grid[0][level]->x;
    int N = solver->grid[0][level]->N;
    double *src_val = solver->grid[SRC][level]->val;
    double *sol_val = solver->grid[SOL][level]->val;

    // Red Sweep
    sol_val[0] = solver->bvp->relax_left_func(sol_val, src_val, x[0], h);
    for (int i = 2; i < N; i += 2)
        sol_val[i] = solver->bvp->relax_middle_func(sol_val, src_val, x[i], h, i);
    sol_val[N] = solver->bvp->relax_right_func(sol_val, src_val, x[N], h);

    // Black Sweep
    for (int i = 1; i < N; i += 2)
        sol_val[i] = solver->bvp->relax_middle_func(sol_val, src_val, x[i], h, i);
}

void bvpsolver_get_residual(BVPSolver *solver, int level)
{
    // Aliases
    double h = solver->grid[0][level]->h;
    double *x = solver->grid[0][level]->x;
    int N = solver->grid[0][level]->N;
    double *src_val = solver->grid[SRC][level]->val;
    double *sol_val = solver->grid[SOL][level]->val;
    double *res_val = solver->grid[RES][level]->val;

    // Residual at the leftmost point
    res_val[0] = solver->bvp->res_left_func(sol_val, src_val, x[0], h);
    // Residual at middle points
    for (int i = 1; i < N; i++)
        res_val[i] = solver->bvp->res_middle_func(sol_val, src_val, x[i], h, i);
    // Residual at the rightmost point
    res_val[N] = solver->bvp->res_left_func(sol_val, src_val, x[N], h);
}


void bvpsolver_multigrid(BVPSolver *solver, int level)
{
    // Pre-Smoothing
    for (int j = 0; j < solver->num_iter1; j++)
        bvpsolver_relax(solver, level);
    if (level > 0) {
        // Error Correction using Coarse Grid
        for (int j = 0; j < solver->num_iter2; j++) {
            bvpsolver_get_residual(solver, level);
            grid_coarsen(solver->grid[RES][level], solver->grid[SRC][level - 1]);
            grid_set_zero(solver->grid[SOL][level - 1]);
            bvpsolver_multigrid(solver, level - 1);
            grid_fine(solver->grid[SOL][level - 1], solver->grid[ERR][level]);
            for (int i = 0; i <= solver->grid[SOL][level]->N; i++) {
                solver->grid[SOL][level]->val[i] += solver->grid[ERR][level]->val[i];
            }
        }

        // Post-Smoothing
        for (int j = 0; j < solver->num_iter3; j++)
            bvpsolver_relax(solver, level);
    }
}

void bvpsolver_get_error(BVPSolver *solver, Grid *solution)
{
    grid_subtract(solution, solver->grid[SOL][solver->n - 1], solver->grid[ERR][solver->n - 1]);
}

// public functions

void bvpsolver_solve(BVPSolver *solver)
{
    bvpsolver_multigrid(solver, solver->n - 1);
}

double bvpsolver_residual_rms(BVPSolver *solver)
{
    bvpsolver_get_residual(solver, solver->n - 1);
    return grid_rms(solver->grid[RES][solver->n - 1]);
}

double bvpsolver_residual_rms_normalized(BVPSolver *solver)
{
    double h = solver->grid[SOL][solver->n - 1]->h;
    return bvpsolver_residual_rms(solver)*h*h;
}

double bvpsolver_error_rms(BVPSolver *solver, Grid *solution)
{
    bvpsolver_get_error(solver, solution);
    return grid_rms(solver->grid[ERR][solver->n - 1]);
}